### ISSUE TITLE
Broker: Reject commands that are based on outdated events

### ISSUE THREAD

**Issue by ThorbenLindhauer** [zeebe-io/zeebe#795]
* Extend `ExecuteCommandRequest` API such that it includes the position of the event that a command builds on
* If this event is not the latest event of a state machine, then reject the command

**Comment by ThorbenLindhauer** [comment]
This requires:

* that responses contain the correct event position (=> #writeEvent must be executed before #executeSideEffects)

When a command response is received, it is not guaranteed that the received event is committed to the partition. That means, the user may assume that this is indeed the latest valid event, which is not true when that event could not be committed. (e.g. use case: user persists event that is returned by command, later uses that event as a reference for another command, this command fails because the event was never committed).

**Comment by ThorbenLindhauer** [comment]
Re the problem with positions of not-yet-committed events: We decided that we will only send the response once the event is committed, so that the returned position is always valid. While this increases the latency for single, isolated requests, we believe that it does not influence the average and worst case under load, because all commands and events share the same log stream, so responses to commands are generally always delayed until prior events have been committed. We should have a look at the performance tests once we have built this.

**Comment by meyerdan** [comment]
@berndruecker  this whould change the behavior we discussed about when the client receives a response for a command. Makes sense for you to watch this

**Comment by menski** [comment]
We decided to go with an approach where we reference the last "command" instead of the resulting event. We created the follow up issues #869 and #870.

@berndruecker I'm sorry but it stays the way it is, responses are send before the event is committed.

**Comment by menski** [comment]
We decided to post pone this issue until we reiterated over workflow/incident processing. At the moment there are cases where events for a stream processor are generated by a different stream processor. This might break or make it difficult to implement the idea from #870. The idea was to keep track of the last record which changed the state of a state machine, e.g. created a new event. If this record is not processed by the stream processor which manages the state machine the update of the state might be difficult. A example would be the resolving of a incident on an input mapping. Resolve incident would create a new activity ready event which means the incident stream processor basically changes the state of the workflow instance processor, which would require a state update of the last record position which changed the state.

As we currently take into consideration to rework the workflow instance stream processing and incident handling/concept we will not continue with this topic until we reiterated on the core concepts. We plan to do this in the near future in preparation of implementing more BPMN concepts/elements.
